src/producer.rs

```diff
@@ -217,34 +217,44 @@ impl Archive {
         }
     }
 
-    pub fn read_in_buffer(&self, name: &str, buf: &mut Vec<u8>) -> bool {
+    pub fn read(&self, name: &str) -> Option<Vec<u8>> {
         match *self.item.borrow_mut() {
             ArchiveType::Zip(ref mut zip) => {
                 let mut zip = zip.borrow_mut();
                 let zipfile = zip.by_name(&name);
                 match zipfile {
                     Ok(mut f) => {
-                        f.read_to_end(buf).expect("Failed to read gcda file");
-                        true
+                        let mut buf = Vec::with_capacity(f.size() as usize + 1);
+                        f.read_to_end(&mut buf).expect("Failed to read gcda file");
+                        Some(buf)
                     }
-                    Err(_) => false,
+                    Err(_) => None,
                 }
             }
-            ArchiveType::Dir(ref dir) => match File::open(dir.join(name)) {
-                Ok(mut f) => {
-                    f.read_to_end(buf).expect("Failed to read gcda file");
-                    true
+            ArchiveType::Dir(ref dir) => {
+                let path = dir.join(name);
+                let metadata = fs::metadata(&path).unwrap();
```

Comment:
Previously we were not failing when we were not able to open a file, we should probably keep that behavior